testcases/kernel/syscalls/recvmmsg/recvmmsg01.c

```diff
@@ -0,0 +1,2389 @@
+/*{
+ * Copyright (c) Google LLC, 2018
+ * SPDX-License-Identifier: GPL-2.0-or-later
+ */
+
+#define _GNU_SOURCE
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <unistd.h>
+#include <string.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <pthread.h>
+#include <assert.h>
+#include <sys/types.h>
+#include <sys/time.h>
+#include <sys/signal.h>
+#include <sys/uio.h>
+#include <sys/file.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <netinet/in.h>
+
+#include "tst_test.h"
+#include "tst_safe_pthread.h"
+
+/*
+ *	Comment out to the #define
+ *		#define	RECVMMSG_USE_SCTP
+ *	to test without SCTP even if it is provided configured in your system.
+ *
+ *	Note that HAVE_NETINET_SCTP_H comes from the:
+ *		#include "tst_test.h"
+ *	so don't mmve this code above that include.
+ */
+
+#ifdef HAVE_NETINET_SCTP_H
+#define	RECVMMSG_USE_SCTP
+#include <netinet/sctp.h>
+#endif
+
+/*}
+ *{	Contants.
+ */
+
+enum { GUARD_CHAR = '+' };		/* guard character to check overwrites*/
+enum { INVALID_SYSCALL_VALUE = -2 };	/* syscalls values are all >= -1 */
+enum { TEST_IP = 0x7F000001 };
+enum { TEST_PORT = 6789 };
+enum { SENDMSG_SLEEP_MS = 10 };
+
+/*
+ * Must be disjoint with errno values and non-zero, value returned by tests
+ * that do multiple recvmsg() calls internally, or that have a non-zero flag,
+ * or that require special treatment (sender needs to sleep).  Those tests
+ * can not be bundled with other tests into a larger recvmmsg(2) multi-
+ * receive message test, some can only be bundled with their own kind.
+ */
+
+enum non_error_errno {
+	TEST_HAS_MULTIPLE_RECVMSG = -1,		/* can't be bundled at all */
+	TEST_CLOEXEC_FLAG = -2,			/* bundle these together */
+	TEST_NEEDS_SLEEP_IN_SENDER = -3,	/* bundle these together */
+};
+
+/*}
+ *{	Typedefs.
+ */
+
+typedef struct timeval		timeval_t;
+typedef struct timespec		timespec_t;
+typedef struct iovec		iovec_t;
+typedef struct msghdr		msghdr_t;
+typedef struct mmsghdr		mmsghdr_t;
+typedef struct sockaddr		sockaddr_t;
+typedef struct sockaddr_in	sockaddr_in_t;
+typedef struct sctp_initmsg	sctp_initmsg_t;
+typedef struct cmsghdr		cmsghdr_t;
+typedef struct call_s		call_t;
+
+typedef void *(*thread_start_t)(void *arg);
+
```

Comment:
No typedefs please, see presentation from Greg HK on why this is wrong:

http://www.kroah.com/linux/talks/ols_2002_kernel_codingstyle_talk/html/mgp00024.html